### DISCUSSION TITLE
SkyPilot for Model Serving?

### DISCUSSION THREAD

**Discussion by sandkoan** [skypilot-org/skypilot#1544] [Q&A]
Has anyone tried to serve ML models through a web server (e.g., Flask, FastAPI) through SkyPilot (for autoscaling, job restart, etc)? Are there any examples detailing how one might go about doing so?

**Comment by concretevitamin** [comment]
Hey @sandkoan, we're aware of a few basic inference use cases on SkyPilot, as well as #994 by @iojw which prototyped using Flask on top of SkyPilot.

If model serving is of interest to you, I'd love to connect to learn more about your use case. Feel free to send me an email or a message through the [community Slack](https://github.com/skypilot-org/skypilot#issues-feature-requests-and-questions)!

**Comment by sandkoan** [comment]
Hey Zongheng,

Hope you're doing well. SkyPilot is brilliant, and quite useful for running jobs in the cloud—it seems to have all the functionality necessary to run something like an Nvidia Triton server that's launched on-demand, and yet, remains long running, in a sense (capable of scaling as necessary). Could you perhaps explicate some of those other ways people have adapted SkyPilot for inference?

Best,
Govind

________________________________

**Comment by concretevitamin** [comment]
@pounde and co have built a model serving server/application called [xView2](https://github.com/plasmo-foss/xview2-ui-backend/tree/sky). Hoping he can chime in more, but off the top of my head these inference requests are heavier computations (10s of minutes) running computer vision models.  @dongreenberg and co have also built [Runhouse](https://github.com/run-house/tutorials) using SkyPilot for all cluster needs, including serving I believe which is slightly less heavier weight(?).

**Comment by kinthaiofficial** [comment]
SkyPilot for model serving is a natural fit — the ability to run on any cloud at spot pricing makes multi-tier agent inference much more economical.

For agent workloads specifically, a few things worth considering in the serving architecture:

**Heterogeneous fleets for tiered routing** — agent traffic has very different latency/quality requirements by task type. Running a fast small model on cheap spot instances (for retrieval/classification) and a capable model on on-demand (for reasoning) cuts blended cost significantly. SkyPilot's cloud-agnostic abstraction is ideal for this since you can mix instance types and clouds based on availability.

**Spot interruption handling** — agents in the middle of a multi-step task can't just restart from scratch when a spot instance is preempted. The serving infrastructure needs to either (a) checkpoint the LLM's KV cache and migrate to another instance, or (b) finish the current request before shutting down. SkyPilot's managed spot could handle (b) but (a) requires coordination with the model server.

**Cost visibility to the orchestrator** — the agent orchestrating the serving layer needs real-time cost data, not just end-of-billing-period summaries. "This instance has been running for 2 hours at $0.50/hr; current task budget is $0.20 remaining" is what lets the orchestrator decide whether to keep the instance or terminate.

**Per-agent cost attribution** — when multiple agents share a serving fleet, you need to track cost per agent request, not just per instance-hour. Without this, cost management is impossible in multi-tenant setups.

The multi-tier inference cost model: https://blog.kinthai.ai/agent-wallet-economic-models-autonomous-agents — the multi-tenancy design that works with shared serving: https://blog.kinthai.ai/openclaw-multi-tenancy-why-vm-per-user-doesnt-scale

Are you considering SkyPilot for development experimentation, or for production serving?

**Comment by pounde** [comment]
I think I can reply here very briefly. Yeah, that's correct that we were running models that would take anywhere between 15 minutes and an hour and a half to run. Essentially we'd point at a large batch of satellite imagery and turn the model loose. I'm not currently working on the project anymore, so I don't know the current state, but that was essentially our use case at the time.
